example/src/main/scala/WSExample.scala

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2022 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import cats.effect._
+import org.http4s.Uri
+import org.http4s.client.websocket.WSFrame
+import org.http4s.client.websocket.WSRequest
+import org.http4s.curl.CurlApp
+import org.http4s.implicits._
+
+object WSExample extends CurlApp.Simple {
+
+  val fastData: Uri = uri"wss://stream.binance.com/ws/btcusdt@aggTrade"
+  val largeData: Uri = uri"wss://stream.binance.com/ws/!ticker@arr"
+  private val websocket = largeData
+
+  def run: IO[Unit] = websocketOrError()
+    .connectHighLevel(WSRequest(websocket))
+    .use { client =>
+      IO.println("ready!") >>
+        client.receiveStream.printlns.compile.drain
+    }
+
+}
+
+object WSEchoExample extends CurlApp.Simple {
+
+  private val echo = uri"wss://ws.postman-echo.com/raw"
+
+  def run: IO[Unit] = websocketOrError()
+    .connectHighLevel(WSRequest(echo))
+    .use { client =>
```

Comment:
Can we adapt this one into a small test for the test suite? The test can be skipped if the curl is not properly configured.

Reply:
I'm working on tests currently, I'll update ASAP

Reply:
I've created a fake server and implemented tests against that server, however I have a problem which might be due to differences in scheduling; when I run websocket examples in ‍`CurlApp`, program runs and gracefully terminates as expected, but when I run it using `CatsEffectSuite` like the other existing suite, it runs and works but does not neither terminate nor timeout!  
I don't have any idea what could be the source of problem! have you encountered this problem before? @armanbilge

Reply:
Yes I have, it probably means that the number of running handles has not returned to `0`, so it keeps looping. Is the handle used for the WS connection being deregistered/canceled somewhere?

https://github.com/http4s/http4s-curl/blob/ea998518ebea920098811579c979b161a03c8bca/curl/src/main/scala/org/http4s/curl/unsafe/CurlExecutorScheduler.scala#L82

Reply:
I guess it does, I added logs on finalization and it seems to work, and above all it works as expected when runs under `CurlApp` but not under `CatsEffectSuite`

Reply:
Just pushed it

Reply:
No, it should work fine.

Reply:
Yeah I'm really confused what's happening.

```
$ uname -a
Linux http4s-http4scurl-pzxb58choln 5.15.0-47-generic #51-Ubuntu SMP Thu Aug 11 07:51:15 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

$ nix --version
nix (Nix) 2.11.0
```

Reply:
No, my shell seems fine. It's a toolchain problem not picking up the right lib for some reason.
```
[http4s-curl]$ nm $LIBRARY_PATH/libcurl.so | grep curl_ws_meta
0000000000071d20 T curl_ws_meta
```

Reply:
Lol, so I added this.
```scala
  } else {
    c.withLinkingOptions(
      s"-L${System.getenv("LIBRARY_PATH")}" +: c.linkingOptions
    )
```
and now I moved on to this 😂 
```
[error] /nix/store/178vvank67pg2ckr5ic5gmdkm3ri72f3-binutils-2.39/bin/ld: /nix/store/9m94s7xxwqa57yj33maxsv03iqa009m3-devshell-dir/lib/libcurl.so: undefined reference to `nghttp2_option_set_no_rfc9113_leading_and_trailing_ws_validation'
```

Reply:
I did a lot of tracing and I'm pretty confident that `curl_easy_cleanup` is called. But it seems that websocket transfers don't change their status on call to `curl_multi_info_read`
So it's not detected for removal from `multi handle` in the scheduler.
However sending a close frame explicitly causes it to be cleaned up, which is not a solution I guess.

Reply:
> But it seems that websocket transfers don't change their status on call to `curl_multi_info_read`

Hmm, that seems like it can be an issue, since that's also used for cleaning up the stored callback.



> However sending a close frame explicitly causes it to be cleaned up, which is not a solution I guess.

Why not? Do you mean server or client sending a close frame? I assume the client always wants to send a close frame, no?

Reply:
I think adding a new `CurlExecutorScheduler.addHandler` that returns a `Resource` can solve the problem.

Reply:
> Why not? Do you mean server or client sending a close frame? I assume the client always wants to send a close frame, no?

I mean the Client, but if it depends on the server to respond it would useless, as interrupted connections won't ever terminate.

Reply:
I added a few tests as you suggested, however it is causing the build to fail currently.